File: leetcode/LC/stack/0084_largest_histogram.py
from typing import List

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []
        maxarea = 0
        i = 0
        while(i < len(heights)):
            if(not len(stack) or heights[stack[-1]] <= heights[i]):
                stack.append(i)
                i += 1
            else:
                tp = stack.pop()
                area = heights[tp]*(i if not len(stack) else (i - stack[-1] -1))
                maxarea = max(area, maxarea)
        
        while(len(stack)):
            tp = stack.pop()
            area = heights[tp]*(i if not len(stack) else (i - stack[-1] -1))
            maxarea = max(area, maxarea)
        return maxarea

File: leetcode/LC1/stack/0084_largest_histogram.cpp
#include <climits>
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack <int> s;
        s.push(-1);
        int maxArea = 0;
        for(int i = 0; i< heights.size(); i++){
            while(s.top() != -1 && heights[s.top()] > heights[i]){
                int height = heights[s.top()];
                s.pop();
                int currentwidth = i - s.top() - 1;
                maxArea = max(maxArea, currentwidth*height);
            }
            s.push(i);
        }
        while(s.top() != -1){
            int height = heights[s.top()];
                s.pop();
                int currentwidth = heights.size() - s.top() - 1;
                maxArea = max(maxArea, currentwidth*height);
        }
        return maxArea;
    }
};
